### PR TITLE
[PI-19] Test that request AVVM given the redeem public key

### PR DESCRIPTION
**Pull request by ntobal**
This PR includes:
- the UI for submitting the redemption
- the implementation of the redemptionKeyToAddress function from the module
- the UTXO request for the address returned
- all the e2e tests for Ada Redemption

Two issues were encountered in the tests: the Force Vended encrypted redemption throws an error in the decryption method and the Paper Vended redemption throws an invalid seed error when redemptionKeyToAddress method is called. These issues will be solved in the following cards:
- https://trello.com/c/dxRSMoLz/51-fix-force-vended-redemption
- https://trello.com/c/xyXihfA9/52-fix-paper-vended-redemption

To run and test this functionality, the module has to be updated running the `setup_cardano_crypto.sh`.

### CODE REVIEW

File: app/api/ada/adaRedemption.js [hunk 1]
```diff
@@ -25,8 +25,9 @@ export async function redeemAda(
   redemptionParams: RedeemAdaParams
 ) : Promise<AdaTransaction> {
   try {
-    const redemptionKey = base64StringToUint8Array(redemptionParams.redemptionCode);
-    // TODO: get address from redemption key: const address = getAddressFromRedemptionKey(redemptionKey);
+    const redemptionKey = Buffer.from(redemptionParams.redemptionCode, 'base64');
+    const addressBuffer = getAddressFromRedemptionKey(redemptionKey);
```
Comment (zuzunker) [inline app/api/ada/adaRedemption.js]: yeah, the problem is that `Redemption.redemptionKeyToAddress` called from `getAddressFromRedemptionKey` implies that `redemptionKey` is the private-key 😬

Comment (zuzunker) [reply app/api/ada/adaRedemption.js]: From discussions with @mirkoAlic we have concluded that required functionality will be to convert **private key** that you get from users (via various input methods) to the avvm address. So I'm not sure if we need to have exposed functionality to convert **public key** to avvm address.

Comment (ntobal) [reply app/api/ada/adaRedemption.js]: We used the public key by mistake when we tried with different inputs, but in the method we are trying to convert the private key.

File: app/api/ada/adaRedemption.js [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+  redemptionCode: string,
+  mnemonic: ?Array<string>,
+  spendingPassword?: string,
+  walletId: string,
+  accountIndex: number
+};
+
+export type RedeemPaperVendedAdaParams = {
+  redemptionCode: string,
+  mnemonic: ?Array<string>,
```
Comment (Juani-Atix) [inline app/api/ada/adaRedemption.js]: Why is the mnemonic optional?

File: app/api/ada/adaRedemption.js [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  mnemonic: Array<string>,
+};
+
+// TODO: return a Promise of AdaTransaction instead of Object
+export async function redeemAda(
+  redemptionParams: RedeemAdaParams
+) : Promise<Object> {
+  const redemptionKey = Buffer.from(redemptionParams.redemptionCode, 'base64');
+  const addressBuffer = getAddressFromRedemptionKey(redemptionKey);
+  const address = bs58.encode(Buffer.from(addressBuffer));
```
Comment (Juani-Atix) [inline app/api/ada/adaRedemption.js]: why do you create a Buffer from addressBuffer? isn't addressBuffer a Buffer already?

File: app/api/ada/adaRedemption.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  return {};
+}
+
+// TODO: return a Promise of AdaTransaction instead of Object
+export async function redeemPaperVendedAda(
+  redemptionParams: RedeemPaperVendedAdaParams
+) : Promise<Object> {
+  const redemptionKey = Buffer.from(redemptionParams.redemptionCode, 'base64');
+  const addressBuffer = getAddressFromRedemptionKey(redemptionKey);
+  const address = bs58.encode(Buffer.from(addressBuffer));
```
Comment (Juani-Atix) [inline app/api/ada/adaRedemption.js]: why do you create a Buffer from addressBuffer? isn't addressBuffer a Buffer already?

File: app/api/ada/lib/utils.js [hunk 5]
```diff
@@ -170,3 +171,11 @@ export function decodeRustTx(rustTxBody: RustRawTxBody): CryptoTransaction {
     throw new Error('Failed to decode a rust tx! Cause: ' + stringifyError(e));
   }
 }
+
+export const encryptPassphrase = (passphrase: ?string) => (
```
Comment (Juani-Atix) [inline app/api/ada/lib/utils.js]: why the type is optional?

File: app/components/wallet/ada-redemption/AdaRedemptionForm.js [hunk 6]
```diff
@@ -369,7 +369,19 @@ export default class AdaRedemptionForm extends Component<Props> {
...
+    this.form.submit({
+      onSuccess: (form) => {
+        const { walletId, shieldedRedemptionKey, spendingPassword } = form.values();
+        this.props.onSubmit({
+          walletId,
+          shieldedRedemptionKey,
+          spendingPassword: spendingPassword || null,
+        });
+      },
+      onError: () => {},
```
Comment (Juani-Atix) [inline app/components/wallet/ada-redemption/AdaRedemptionForm.js]: Unhandle error here

File: app/containers/wallet/AdaRedemptionPage.js [hunk 7]
```diff
@@ -15,6 +15,18 @@ import environment from '../../environment';
 export default class AdaRedemptionPage extends Component<InjectedProps> {
   static defaultProps = { actions: null, stores: null };
 
+  onSubmit = (values: { walletId: string, spendingPassword: ?string }) => {
```
Comment (Juani-Atix) [inline app/containers/wallet/AdaRedemptionPage.js]: why is spendingPassword type optional?

File: app/stores/ada/AdaRedemptionStore.js [hunk 8]
```diff
@@ -181,6 +196,87 @@ export default class AdaRedemptionStore extends Store {
...
+    walletId: string,
+    spendingPassword: ?string,
+  }) => {
+
+    runInAction(() => { this.walletId = walletId; });
+
+    // Since there's no support for multiwallet yet, the only account index present in localStorage
+    // is used.
+    const accountData = getSingleCryptoAccount();
+    const accountIndex = accountData.account;
```
Comment (Juani-Atix) [inline app/stores/ada/AdaRedemptionStore.js]: why don't you use object destructuring?

File: app/api/ada/index.js [hunk 9]
```diff
@@ -615,6 +618,50 @@ export default class AdaApi {
...
+    const spendingPassword = passwordString ? encryptPassphrase(passwordString) : '';
+    try {
+      const transaction = await redeemAda(
+        { ...request, spendingPassword }
+      );
+      Logger.debug('AdaApi::redeemAda success');
+      // TODO: call the following function once the tx is broadcasted:
+      // return _createTransactionFromServerData(transaction);
+    } catch (error) {
+      Logger.debug('AdaApi::redeemAda error: ', error);
```
Comment (Juani-Atix) [inline app/api/ada/index.js]: Logger.debug only takes 1 argument

File: app/api/ada/adaRedemption.js [hunk 10]
```diff
@@ -0,0 +1,46 @@
+// @flow
+
+import { getAddressFromRedemptionKey } from './lib/cardanoCrypto/cryptoRedemption';
+import bs58 from 'bs58';
+import { getUTXOsForAddresses } from './lib/yoroi-backend-api';
+
+export type RedeemAdaParams = {
+  redemptionCode: string,
+  mnemonic: ?Array<string>,
```
Comment (Juani-Atix) [inline app/api/ada/adaRedemption.js]: its `mnemonics`

**Pull request merged**